### PR TITLE
When a tensor value produced by accelerator is to be sent to host, if all its users are on the host (similar to the case of a result tensor from a TF program), stop issuing an accelerator-to-host copy warning for it.

### PR DESCRIPTION
**Pull request by mhong**
One use case is outfeeding the loss value for logging/debugging within a TF
training loop, where such data transfer is async with respect to TF computation,
and is not expected to hurt TF performance.

The motivating test case for this PR is the following:
```swift
public func SR8313(_ n: Int) -> Tensor<Double> {
  var t = Tensor(1)
  var i = 0
  while i < n {
    t += 1
    i += 1
  }
  return t
}
```

The marked SIL code from TFPartition pass is:

```
---- ANALYSIS STATE FOR FUNCTION $S8test_tmp6SR8313y10TensorFlow0D0VySdGSiF ----------
Tensor start point:   %5 = integer_literal $Builtin.Int64, 0          // user: %11
Tensor end point:   %27 = struct $Tensor<Double> (%26 : $TensorHandle<Double>) // user: %28
SIL with markings:

bb0:
  %0 = argument of bb0 : $Int                       // users: %3, %1
    debug_value %0 : $Int, let, name "n", argno 1   // id: %1
    %2 = integer_literal $Builtin.Int64, 0          // user: %4
[Arg]   %3 = struct_extract %0 : $Int, #Int._value      // users: %4, %22
[Arg]   %4 = builtin "cmp_slt_Int64"(%2 : $Builtin.Int64, %3 : $Builtin.Int64) : $Builtin.Int1 // user: %7
[Copy]    %5 = integer_literal $Builtin.Int64, 0          // user: %11
[Move]    %6 = graph_op "Const"() {dtype: $Double, value$tensor: f64 0x3FF0000000000000 /* 1 */, __device: "ALL_DEVICES"} : $TensorHandle<Double> // users: %8, %11
[Copy]    cond_br %4, bb2, bb1                            // id: %7

bb1:
    br bb6(%6 : $TensorHandle<Double>)              // id: %8

bb2:
[Copy]    %9 = integer_literal $Builtin.Int64, 1          // user: %18
    %10 = integer_literal $Builtin.Int1, -1         // user: %18
[Copy]    br bb3(%6 : $TensorHandle<Double>, %5 : $Builtin.Int64) // id: %11

bb3:
[Move]  %12 = argument of bb3 : $TensorHandle<Double>     // users: %17, %15
[Copy]  %13 = argument of bb3 : $Builtin.Int64            // user: %18
[Move]    %14 = graph_op "Const"() {dtype: $Double, value$tensor: f64 0x3FF0000000000000 /* 1 */, __device: "ALL_DEVICES"} : $TensorHandle<Double> // users: %16, %15
[Move]    %15 = graph_op "Add,i,i"(%12 : $TensorHandle<Double>, %14 : $TensorHandle<Double>) {T: $Double, __device: "/device:CPU:0"} : $TensorHandle<Double> // users: %24, %25
    strong_release %14 : $TensorHandle<Double>      // id: %16
    strong_release %12 : $TensorHandle<Double>      // id: %17
[Copy]    %18 = builtin "sadd_with_overflow_Int64"(%13 : $Builtin.Int64, %9 : $Builtin.Int64, %10 : $Builtin.Int1) : $(Builtin.Int64, Builtin.Int1) // users: %20, %19
[Copy]    %19 = tuple_extract %18 : $(Builtin.Int64, Builtin.Int1), 0 // users: %25, %22
    %20 = tuple_extract %18 : $(Builtin.Int64, Builtin.Int1), 1 // user: %21
    cond_fail %20 : $Builtin.Int1                   // id: %21
[Copy]    %22 = builtin "cmp_slt_Int64"(%19 : $Builtin.Int64, %3 : $Builtin.Int64) : $Builtin.Int1 // user: %23
[Copy]    cond_br %22, bb5, bb4                           // id: %23

bb4:
    br bb6(%15 : $TensorHandle<Double>)             // id: %24

bb5:
[Copy]    br bb3(%15 : $TensorHandle<Double>, %19 : $Builtin.Int64) // id: %25

bb6:
  %26 = argument of bb6 : $TensorHandle<Double>     // user: %27
    %27 = struct $Tensor<Double> (%26 : $TensorHandle<Double>) // user: %28
    return %27 : $Tensor<Double>                    // id: %28
---- END OF ANALYSIS STATE FOR FUNCTION ----------
```

From the above, `br bb6(%6 : $TensorHandle<Double>)` and `br bb6(%15 :
$TensorHandle<Double>)` are not marked for accelerator, so they triggered
accel-to-host copy warnings for `%6` and `%15` before this patch. However, these
tensors effectively serve as result tensors of the program (this point becomes
more clear, if we add a `t += 0.0` right after the while loop in the above
test case, where the resulting tensor would become a result tensor of the TF
program), so issuing copy warnings for them is not very useful.

Deleted the `scalarToHost()` test case from
`diagnostics_scalar_transfers.swift`, because it's difficult to construct a test
case where we'd still warn on copying a scalar value from accelerator to host,
but the other test `scalarToAccelerator()` provided coverage for the compiler
flag `-tf-warn-scalar-transfer`.

Resolves [SR-8313](https://bugs.swift.org/browse/SR-8313).

### CODE REVIEW

**Comment by mhong** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
You have convinced me that eliminating warnings for transfers that are async with respect to the TF computation is probably a good idea.

But whether the transferred value gets used in graph ops seems orthogonal to whether it's async wrt the computation. For example, here's a transfer that's async even though the value gets use in graph ops:
```
// weights and inputs are on accelerator
loop {
  let predictions = predict(weights, inputs) // predict happens on accelerator
  print(predictions) // transfers predictions to host
  let gradients = gradient(weights, inputs, predictions) // gradient happens on accelerator
  weights -= gradients
}
```

Therefore, what if we just eliminate all transfer-to-host warnings? This will eliminate warnings for transfers that are async with respect to the computation. Transfers that block the computation should still produce warnings, because they'll emit copy-to-accelerator warnings!

**Comment by mhong** [comment]
Discussed a bit more in person. I like the simplicity of only warning on host->TF transfers!

One potential concern is that TF->host transfers within a loop can still potentially slow down TF performance. e.g.
```swift
while ... {
  weights -= tf_computation(...)
  print(weights)
}
```

Since weights keep being updated, if the transfer is not fast enough, and TF does not have enough RAM to buffer the older copies of `weights` to be sent to host, that can block TF computation. This concern remains valid also in the case of `weights` being functionalized -- where each loop iteration produces a new copy.

Marc made a fair point that whether this would actually impact TF performance depends on the runtime context -- how fast the network is, how much TF RAM, and how large is `weights`, etc. So it's not clear that issuing a compile-time warning is always the most appropriate thing to do.

As the next step, Marc will write up a doc to seek more team/community input, and I'll hold off on this PR.

**Comment by lattner** [comment]
I don't think that this patch is the right direction, but I think that you're on to something here.  My objection to this patch in particular is that there is no way to know whether something is "async".  I think you're trying to allow things like:

let x = tensorstuff()
print(x)

But it is very simple for this to become synchronous w.r.t. the accelerator:

var x = tensorstuff()
x = sin(x)   // copy to host, libm sin(), copy back to accel
tensorstuff(x)


I think that sends of scalar values from the host to the accelerator should not be a warning: this should be implementable with an efficient broadcast to all the replicas, and is the vastly most common case for unstructured control flow (where we send out bools :).

It's also worth pointing out that the case in the testcase only happens because the function is marked public, which isn't what normal models will do.  We do this in the testsuite because it is a simple way to force a deabstraction/partitioning scope.
